server-utils/src/suspendable_stream.rs

```diff
@@ -0,0 +1,63 @@
+use std::time::{Duration, Instant};
+use tokio::timer::Delay;
+use std::io;
+use tokio::prelude::*;
+
+/// A wrapper type for tokio::Incoming that stops accepting new connections
+/// for a specified amount of time once `ulimit` is reached
+pub struct SuspendableStream<S> {
+	stream: S,
+	delay: Duration,
+	timeout: Option<Delay>,
+}
+
+impl<S> SuspendableStream<S> {
+	/// construct a new Suspendable stream, given tokio::Incoming
+	/// and the amount of time to pause for.
+	pub fn new(stream: S) -> Self {
+		SuspendableStream {
+			stream,
+			delay: Duration::from_millis(500),
+			timeout: None,
+		}
+	}
+}
+
+impl<S, I> Stream for SuspendableStream<S>
+	where S: Stream<Item=I, Error=io::Error>
+{
+	type Item = I;
+	type Error = ();
+
+	fn poll(&mut self) -> Result<Async<Option<Self::Item>>, ()> {
+		let mut recovered = None;
+		if let Some(mut timeout) = self.timeout.take() {
+			match timeout.poll() {
+				Ok(Async::Ready(_)) => {
+					recovered = Some(());
+				}
+				Ok(Async::NotReady) => {
+					self.timeout = Some(timeout);
+					return Ok(Async::NotReady)
+				},
+				Err(_) => unreachable!("Polling a delay shouldn't yield any errors")
+			}
+		}
+
+		loop {
+			match self.stream.poll() {
+				Ok(item) => return Ok(item),
+				Err(err) => {
```

Comment:
I wonder if we should `match` on the underlying `stream`'s possible errors here and decide which ones we should suspend for and which ones are fatal? I'm not sure about this but aren't there other error conditions where we do not want to suspend-resume forever?

Reply:
Maybe you look at the following commit: https://github.com/hyperium/hyper/commit/68458cde57a20f4b3c9c306eaf9801189262e0a6
That introduced `sleep_on_error` in hyper. That's similar to what you are want to achieve here. 
Look into the `connection_error` function, this defines errors that should not end in sleeping.